cli/src/commands/create/comments.rs

```diff
@@ -102,112 +158,239 @@ pub fn create(client: &Client, args: &CreateCommentsArgs) -> Result<()> {
+    let mut comments_to_put = Vec::with_capacity(batch_size);
+    let mut comments_to_sync = Vec::new();
+    let mut annotations = Vec::new();
 
-            if dataset_name.is_some() && new_comment.has_annotations() {
-                with_annotations.push((
-                    CommentUid(format!("{}.{}", source.id.0, new_comment.comment.id.0)),
-                    new_comment.clone(),
-                ));
-            }
-            batch.push(new_comment.comment);
+    // if --overwrite, everything will go to comments_to_sync, so put the default capacity there.
+    if overwrite {
+        std::mem::swap(&mut comments_to_put, &mut comments_to_sync)
+    }
+
+    let mut is_duplicate_id: Box<dyn FnMut(String) -> bool> = if allow_duplicates && overwrite {
+        Box::new(|_| {
+            panic!("should never check for duplicates with both --allow-duplicates and --overwrite")
+        })
+    } else {
+        let mut seen = HashSet::new();
+        Box::new(move |id| !seen.insert(id))
+    };
+
+    for read_comment_result in read_comments_iter(comments, Some(statistics)) {
+        let new_comment = read_comment_result?;
+
+        if dataset_name.is_some() && new_comment.has_annotations() {
+            annotations.push((
+                CommentUid(format!("{}.{}", source.id.0, new_comment.comment.id.0)),
+                new_comment.labelling,
+                new_comment.entities,
+            ));
         }
 
-        if batch.len() == batch_size || (!batch.is_empty() && eof) {
-            // Upload comments
-            let result = client
-                .sync_comments(&source.full_name(), &batch)
-                .chain_err(|| ErrorKind::Client("Could not upload batch of comments".into()))?;
-            let num_uploaded = batch.len();
-            statistics.add_comments(StatisticsUpdate {
-                uploaded: num_uploaded,
-                new: result.new,
-                updated: result.updated,
-                unchanged: result.unchanged,
-            });
-            batch.clear();
-
-            // Upload annotations
-            if let Some(dataset_name) = dataset_name.as_ref() {
-                for (comment_uid, new_comment) in with_annotations.iter() {
-                    client
-                        .update_labelling(
-                            &dataset_name,
-                            comment_uid,
-                            new_comment.labelling.as_ref(),
-                            new_comment.entities.as_ref(),
-                        )
-                        .chain_err(|| {
-                            ErrorKind::Client("Could not update labelling for comment".into())
-                        })?;
-                    statistics.add_annotation();
-                }
-                with_annotations.clear();
-            }
+        if overwrite || (allow_duplicates && is_duplicate_id(new_comment.comment.id.0.clone())) {
+            comments_to_sync.push(new_comment.comment);
```

Comment:
It would be nice, but the downside is that `--allow-duplicates` + `--overwrite` combined would have to start counting duplicates. I understood the goal was to avoid memory growth with that combination.

Reply:
Yes, good point